google/cloud/storage/internal/openssl_util.h

```diff
@@ -0,0 +1,245 @@
+            static_cast<BIO*>(pem_buffer.get()),
+            nullptr,  // EVP_PKEY **x
+            nullptr,  // pem_password_cb *cb -- a custom callback.
+            // void *u -- this represents the password for the PEM (only
+            // applicable for formats such as PKCS12 (.p12 files) that use
+            // a password, which we don't currently support.
+            nullptr),
+        &EVP_PKEY_free);
+    if (not(private_key)) {
+      handle_openssl_failure("Could not parse PEM to get private key.");
+    }
+
+    const int DIGEST_SIGN_SUCCESS_CODE = 1;
+    if (DIGEST_SIGN_SUCCESS_CODE !=
+        EVP_DigestSignInit(static_cast<EVP_MD_CTX*>(digest_ctx.get()),
+                           nullptr,  // EVP_PKEY_CTX **pctx
+                           digest_type,
+                           nullptr,  // ENGINE *e
+                           static_cast<EVP_PKEY*>(private_key.get()))) {
+      handle_openssl_failure("Could not initialize PEM digest.");
+    }
+
+    if (DIGEST_SIGN_SUCCESS_CODE !=
+        EVP_DigestSignUpdate(static_cast<EVP_MD_CTX*>(digest_ctx.get()),
+                             str.c_str(), str.length())) {
+      handle_openssl_failure("Could not update PEM digest.");
+    }
+
+    std::size_t signed_str_size = 0;
+    // Calling this method with a nullptr buffer will populate our size var
+    // with, the resulting buffer's size. This allows us to then call it again,
+    // with the correct buffer and size, which actually populates the buffer.
+    if (DIGEST_SIGN_SUCCESS_CODE !=
+        EVP_DigestSignFinal(static_cast<EVP_MD_CTX*>(digest_ctx.get()),
+                            nullptr,  // unsigned char *sig
+                            &signed_str_size)) {
+      handle_openssl_failure("Could not finalize PEM digest (1/2).");
+    }
+
+    std::vector<unsigned char> signed_str(signed_str_size);
+    if (DIGEST_SIGN_SUCCESS_CODE !=
+        EVP_DigestSignFinal(static_cast<EVP_MD_CTX*>(digest_ctx.get()),
+                            signed_str.data(), &signed_str_size)) {
+      handle_openssl_failure("Could not finalize PEM digest (2/2).");
+    }
+
+    return std::string(signed_str.begin(),
+                       signed_str.begin() + signed_str_size);
+  }
+
+  /**
+   * Return a Base64-encoded version of the given a string, using the URL- and
+   * filesystem-safe alphabet, making these adjustments:
+   * -  Replace '+' with '-'
+   * -  Replace '/' with '_'
+   * -  Right-trim '=' characters
+   */
+  static std::string UrlsafeBase64Encode(std::string const& str) {
+    std::string b64str = Base64Encode(str);
+    std::replace(b64str.begin(), b64str.end(), '+', '-');
+    std::replace(b64str.begin(), b64str.end(), '/', '_');
+    RightTrim(b64str, '=');
+    return b64str;
+  }
+
+ private:
+  static std::unique_ptr<BIO, decltype(&BIO_free_all)>
+  MakeBioChainForBase64Transcoding() {
+    std::ostringstream err_builder;
+    auto base64_io = std::unique_ptr<BIO, decltype(&BIO_free)>(
```

Comment:
Would `using BioUniquePtr = std::unique_ptr<BIO, decltype(&BIO_free)>` make the code more readable?  I do not know, so feel free to ignore this.

Reply:
I actually thought about that, but wanted the deleter function to be explicit and directly beside each BIO declaration.  The difference between BIO_free and BIO_free_all is important (i.e. calling BIO_free on a BIO that was added to a chain that was freed up with BIO_free_all will result in a crash from a double free).